daemon/cmd/ipam.go

```diff
@@ -283,6 +284,89 @@ func (d *Daemon) allocateHealthIPs() error {
 	return nil
 }
 
+func (d *Daemon) allocateIngressIPs() error {
+	bootstrapStats.ingressIPAM.Start()
+	if option.Config.EnableEnvoyConfig {
```

Comment:
Something else that Thomas raised for discussion out-of-band is that allocating new IPs is problematic for chaining modes where another plugin handles IPAM. In those cases there's not always a guarantee that Cilium can hand an IP out. That said, we don't have any specific mode for this today. Probably we should look into adding an `--ipam=disabled` option and just forbid features like this in chaining mode, until we figure out another way to solve this. (NB: I don't see this as a blocker to merge this PR, it's sort-of an existing problem that should be handled separately).

One plausible alternative is just to say that in chaining mode, all ingress traffic shows up as `remote-node`. We'd have to ensure that when the Envoy on a node directs traffic towards a backend on the current node, that it hands the `remote-node` identity down to Cilium's eBPF programs (and that this identity doesn't get replaced by the `OUTPUT` chain iptables rule today which identifies traffic as "from `host`").

Reply:
I tried to avoid the `remote-node`, which I think is a stop-gap solution at best.

Maybe we could add dummy ingress endpoints to all the nodes and make the agent find those based on some annotation and then just use the IPs allocated for them? That should work in chained modes as well?

Reply:
yeah agree that the policy posture of tying it to the same L3 IP / identity with `remote-node` is not ideal.

Dummy ingress endpoints seems like a reasonable way to get IPAM in that situation yeah. Then in future also if we look towards selectively choosing which nodes the ingress is exposed on (eg for massive clusters), the affinity for those dummy ingress endpoints would be the way to ensure that you only allocate IPs on the relevant nodes.

There's some usability quirks there - for instance, how do we name this and communicate it such that it's clear to the user that this is just a holder for an IP address. The traffic woludn't actually go via that pod, and you can't do anything with that pod, but it's necessary to implement ingress, we just steal its IP for reuse there. One idea is to name it the `cilium-ingress-drone`, fits the bee theme and drones have just one job in the colony, in this case allowing ingress to work.

I also wonder whether it should actually take on specific labels instead of annotations, eg `io.cilium/app=ingress`. This way it's harder to mess with due to the way that k8s RBAC works.

Reply:
^ If we do this, I think we could actually provide the full dummy spec in the Helm charts and deploy it for the user if they enable the ingress features.

Reply:
I'll experiment with the dummy Endpoint as an alternative to doing IPAM directly once I have time.

Reply:
@joestringer Looked into this and got the `cilium-ingress-drone` pod added to our Helm charts, updated `cilium-cli` (locally) to deploy it (`cilium-cli` does not simply apply the helm manifest as such, but picks and chooses parts of it), and updated Cilium agent to notice when such a pod is added or restored. Before proceeding with testing I found out that we do not support L7 in chaining CNI modes, so L7 LB/ Ingress would likely not work either.

Given the above I think we need to postpone CNI chaining mode support until after 1.12 release, if ever.

If so, this PR should be good to merge as-is. Also, I came to the conclusion that we probably would want the agent bootstrap to depend on a pod deployment only when needed due to the unavailability of IPAM, as it adds a new failure mode (the dummy pod does not get deployed soon enough and Cilium agent bootstrap fails). Hence the CNI chaining mode support would be additional work on top of the approach in this PR, not a replacement for it.